File: frontend/src/types/borrow.ts
// 借阅类型，契约见 docs/system-design.md 8.6
import type { BookBriefOut } from './book'
import type { ReaderBriefOut } from './reader'
import type { UserBriefOut } from './user'

export type BorrowStatus = 'borrowed' | 'overdue' | 'returned'

export interface BorrowOut {
  id: number
  book: BookBriefOut
  reader: ReaderBriefOut
  borrowed_by: UserBriefOut
  borrowed_at: string
  due_date: string
  returned_at: string | null
  returned_by: UserBriefOut | null
  status: BorrowStatus
  notes: string | null
}

export interface BorrowCreate {
  book_id: number
  reader_id: number
  due_date?: string
  notes?: string
}

export interface BorrowQuery {
  page?: number
  page_size?: number
  status?: BorrowStatus
  book_id?: number
  reader_id?: number
  due_before?: string
}

File: backend/app/schemas/borrow.py
from datetime import date, datetime
from typing import Literal

from pydantic import Field, field_validator

from backend.app.schemas.book import BookBriefOut
from backend.app.schemas.common import SchemaBase
from backend.app.schemas.reader import ReaderBriefOut
from backend.app.schemas.user import UserBriefOut


class BorrowCreate(SchemaBase):
    book_id: int = Field(gt=0)
    reader_id: int = Field(gt=0)
    due_date: date | None = None
    notes: str | None = Field(default=None, max_length=500)

    @field_validator("notes", mode="before")
    @classmethod
    def strip_notes(cls, value):
        return value.strip() if isinstance(value, str) else value


class BorrowOut(SchemaBase):
    id: int
    book: BookBriefOut
    reader: ReaderBriefOut
    borrowed_by: UserBriefOut
    borrowed_at: datetime
    due_date: date
    returned_at: datetime | None
    returned_by: UserBriefOut | None
    status: Literal["borrowed", "overdue", "returned"]
    notes: str | None
